### ISSUE TITLE
New Feature Todo List

### ISSUE THREAD

**Issue by Zet0rz** [Zet0rz/nzombies#304]
This is mainly a list so we can keep track of what new things we need to make both before, and as possible post-workshop release. This is the next bits coming, potentially as part of the upcoming large update.
### Workshop Release
- Hammer .fgd _(With "Map Config" if needed, similar to Gamemode Extensions)_
- ~~Map/Config Switch system (for dedicated servers)~~ _(Finished in dev branch)_
- ~~New Special Weapons hooks (Likely use [PlayerButtonDown](http://wiki.garrysmod.com/page/GM/PlayerButtonDown) which is predicted)~~ _(Finished in dev branch)_
- Many more hooks
- ~~Widow's Wine~~ _(Finished in dev branch)_
- Gobblegum (Possible?)
- Traps
- ~~Death Machine~~ _(Finished in dev branch)_
- ~~Zombie Shield~~ _(Finished locally)_
- ~~Buildable Tables~~ _(Finished in dev branch)_

**Comment by Zet0rz** [comment]
![](https://pbs.twimg.com/media/CsK--q-WIAAHjNA.jpg)
I have currently got the Death Machine with c_arms. Its firing animation is still broken (1 frame short), but I should be able to quickly make that. Powerup drop also in the materials I got from Raven to make this.

(Don't worry, it won't have ammo as shown here, this is just me testing the model itself in sandbox)

**Comment by minin43** [comment]
The ammo was infinite until time ran out, no?

**Comment by Zet0rz** [comment]
> The ammo was infinite until time ran out, no?

Yes, in CoD it is and in nZombies it will be.

**Comment by Lucius1202** [comment]
The whole list is fantastic! I give a place of honor to the traps, which are essential in the gameplay and I find that implement them is great!

I take this opportunity to give an idea, in creative place three props like the esteregg (chosen by the user) that activate special doors or special weapons, every time you turn a prop this will disappear and when activated all three, a special flag is activated. I hope I explained.

**Comment by Zet0rz** [comment]
![](http://i.imgur.com/VEhlo9e.png)

Death Machine is now a fully-functional powerup in-game (with the exception of the model itself which will come with the update to the content pack). It also brings some new functionality, such as the ability to "cancel" powerups on players, and the ability for special weapons to be exceptionally holstered (Death Machine uses this so you can holster it, cancelling the powerup)

**Comment by Zet0rz** [comment]
![](http://i.imgur.com/LUIJctk.png)

Wouldn't you look at that! Time to actually get it working though :D

**Comment by lolleko** [comment]
Traps are in progress aswell. Base code and a simple turret are already functional. Still have to do the tesla trap and serialisation. I will have some spare time soon, to finish it.

**Comment by Zet0rz** [comment]
Okay Widow's Wine now freezes zombies in place with your grenades and modifies them to become sticky. It doesn't change the model to the WW grenades, I'm not sure whether to try and get them or just stick with normal grenades.

Now I just need to trigger the same effect on being hurt, as well as a web aura effect and web explosion effect.

**Comment by Lucius1202** [comment]
You guys are great, seriously.

**Comment by Zet0rz** [comment]
![](http://i.imgur.com/JXiSFI6.png)

I'm actually kinda satisfied with how nice the normal glass shot decal texture looks as webs

**Comment by Zet0rz** [comment]
Got the perk icons, perk bottle materials, death machine models, and perk models in the local content pack now. Going to get the red default trap handle as well as the tesla coil thingies that emit the electricity for the traps, then update the github as well as the content pack.

**Comment by Zet0rz** [comment]
![](http://i.imgur.com/uN99VYL.png)

Added a little requested feature as well. Should be easier than searching for models specifically if you know where they come from. These can of course also be added to categories in the main spawnlist by right-clicking.

**Comment by minin43** [comment]
Damn son, you're making this so incredibly easy for config creators

**Comment by Zet0rz** [comment]
It's supposed to be ;)

**Comment by Zet0rz** [comment]
![](http://i.imgur.com/x5XQadx.jpg)

Spooky stuff man, this round won't be easy!

**Comment by func-octo** [comment]
I thought the infinity symbol could use some theming since it looks too smooth for nZombies, so I madea modified one that I think fits the font.
![infinity-nzombies](https://cloud.githubusercontent.com/assets/10675391/18784619/45d72a7a-8161-11e6-8ea7-facb8b9d27cc.png)

Should be the normal size and fit normally.

**Comment by Zet0rz** [comment]
Yeah I was gonna remake it too, just used it temporarily really. I like yours, but I might make it a bit more drippy/blood-like

**Comment by Zet0rz** [comment]
What do you guys think of a door buying effect like this?
[http://i.imgur.com/WuCbcgW.gifv](http://i.imgur.com/WuCbcgW.gifv)
They shake for 0.3 seconds, then fly up along with some lightning field effect like in the game. The arcs could probably use a bit of adjustment based on the height of the entity. If I could merge the effect for groups of entities, that'd be even better. The effect uses the entity's OBB to determine where to start somewhere in the top and then arc down to the bottom.

**Comment by minin43** [comment]
Yes I've been waiting for this.

**Comment by func-octo** [comment]
Any way to nab the cow mangler disintegration effect from tf2 (or just the pulse rifle secondary) and dissolve the models? Personally, that'd look better.

**Comment by lolleko** [comment]
> Any way to nab the cow mangler disintegration effect from tf2 (or just the pulse rifle secondary) and dissolve the models? Personally, that'd look better.

A dissolve effect would be cool especially for ragdolls.

We could try to get https://developer.valvesoftware.com/wiki/Env_entity_dissolver working in gmod. 

Setting the dmg flag [DMG_DISSOLVE](https://github.com/ValveSoftware/source-sdk-2013/blob/0d8dceea4310fde5706b3ce1c70609d72a38efdf/mp/src/game/shared/shareddefs.h#L433) might also work, Although i have no idea if you can apply that to props.

 TL;DR I will try it but i dont have high hopes tbh.

**Comment by Zet0rz** [comment]
The dissolving effect exists in gmod but I'm not sure how to trigger it. Surely there's a way though.

**Comment by func-octo** [comment]
Awesome :) Good luck with the dissolving entity! If you get this to work, keep the electrical arcs because it would look great in combination with the dissolving effect. If it isn't too resource heavy you could make the door item(s) vibrate while they dissolve :D

**Comment by Zet0rz** [comment]
Oh you meant for the doors? They won't dissolve, they just fly up and disappear. They also already vibrate, it's just not so easy to see on that gif. It's done clientside so it shouldn't be too resource intensive.

**Comment by func-octo** [comment]
Alright. Do pulse rifle secondary shots affect the zomies or are they nextbot type bots?

**Comment by Zet0rz** [comment]
What do you guys think of this?

![](http://i.imgur.com/6uqkkim.png)

**Comment by Roachisserie** [comment]
With regards to combine balls. Its easy enough to run a findinsphere to fire an input to cballs to make them explode.

Then you just deal dissolve damage.

**Comment by Roachisserie** [comment]
Oh and Zet0r. You can trigger the dissolving effect by dealing dissolve damage.

I'm not sure how to access the effect itself though.

**Comment by minin43** [comment]
The "dissolve" effect from HL2 would be the perfect source-ification of the gamemode (much like how the melee is a crowbar and not a knife). If the dissolve effect could be added to the props (and not doors) that would be awesome.

**Comment by func-octo** [comment]
I like that beam a lot more tbh. Looks smoother. You could even make it say nZombies or Garry's Mod down it, but idk if that'd look good.

**Comment by func-octo** [comment]
I just thought of something that may be problematic. If you have a skyscraper type level, and you open a door downstairs, could it crush someone standing on the above floor?

**Comment by lolleko** [comment]
My guess is that since the flying up objects are clientside only there will be no collisions with these objects.

**Comment by func-octo** [comment]
Lol, that'd be funny actually, just make the map even more dangerous. You could be right, but the door would have to be open for everyone, so they could somehow crush themselves?

**Comment by Zet0rz** [comment]
It's completely clientsided. It notices when it's no longer locked yet still not nodrawn, notes that time down and when 0.3 seconds has passed, sets a position that flies up with a constant speed for 0.5 seconds. It then in its Draw function sets its RenderOrigin (the position from which the model is rendered) to this position and adds a little shake by adding a randomized vector to it. 

The entity itself never moves, only the position from which its model is drawn.

**Comment by func-octo** [comment]
Could you make it so it would be easy for us to make a custom door buy effect in general or for map scripts?

**Comment by Zet0rz** [comment]
Probably not no

**Comment by Zet0rz** [comment]
I got the can't reload bug fixed! :D

**Comment by Zet0rz** [comment]
Content pack's gonna be split into 2 parts now! With the additions, the single content pack was simply too big to be compressed by gmpublish xD

**Comment by Zet0rz** [comment]
Content packs updated! Be sure to get both [Part 1](http://steamcommunity.com/sharedfiles/filedetails/?id=675138912) and [Part 2](http://steamcommunity.com/workshop/filedetails/?id=777195612).

**Comment by lolleko** [comment]
Why multiple parts?

**Comment by Zet0rz** [comment]
`Compressing [241.8 mb] -> Failed!`

**Comment by lolleko** [comment]
Right probably over 100mb after compressing.

**Comment by Zet0rz** [comment]
It was before 128mb after compression, 210mb before, don't think it is that limit, but it's definitely because it was too big. Should also prevent the crashes people are having from downloading too big addons.

**Comment by lolleko** [comment]
100mb was valves size cap for workshop files, maybe they upped it tho.

**Comment by minin43** [comment]
I have uploaded workshop items with over 200 mb of data after being compressed, yet it's failed with 100 mb pre-compressed. Don't trust any number.

**Comment by Zet0rz** [comment]
I was gonna make a Zombie Shield as I wanted it to be usable in one of my easter eggs. I thought I should make it by combining 3 gmod-themed items, so I combined a fence, a radiator, and a tic-tac-toe-post to create this:
![](http://i.imgur.com/715uX0L.png)

What do you guys think? It'll be buildable (but require a map script to do so), after which I want it to parent to the back spine bone like the picture.

**Comment by Zet0rz** [comment]
![](https://fat.gfycat.com/ExaltedLeanGemsbok.gif)
This might actually scare me

Fence is oddly transparent for some reason though ...

**Comment by func-octo** [comment]
If there was a zombie invulnerable to fire from the back, that would scare me as well xD

**Comment by Zet0rz** [comment]
![](https://thumbs.gfycat.com/BigPleasantArchaeopteryx-size_restricted.gif)
https://gfycat.com/BigPleasantArchaeopteryx

**Comment by Zet0rz** [comment]
![](http://i.imgur.com/PimTr0M.jpg)

**Comment by func-octo** [comment]
There should be a shield that reduces fire damage 60% for 30 seconds, but you have a 10% reduced fire rate and 20% longer reload time during this, so it _could_ be easier to beat the panzer with fewer players.

**Comment by Zet0rz** [comment]
Wow that was quick, it's fully working. Correctly blocks damage depending on angle and whether you hold it or it's on your back, and it breaks and has sounds and everything. Now I just gotta make the buildable table that can be scripted in with a map script (tool might come later).

> There should be a shield that reduces fire damage 60% for 30 seconds, but you have a reduced fire rate and reload time during this, so it could be easier to beat the panzer with fewer players.

There's only this one shield. Following the CoD style, it's themed after gmod, so it's made of a radiator, a fence, and a tic-tac-toe post from a playground. It does nothing but prevent damage if it was dealt from behind when on back or from in front when equipped. It can also melee kill 1-hit up until round 14.

**Comment by minin43** [comment]
> So it could be easier to beat the panzer

The Panzer should always be difficult to kill, no matter how many people are on. It should scale in hp based on alive players when it spawns

Also, don't make the shield overpowered. What's the point to getting the bowie knife if you could instead get this?

**Comment by Zet0rz** [comment]
> Also, don't make the shield overpowered. What's the point to getting the bowie knife if you could instead get this?

I just copy-pasted damage from CoD (actually not, in CoD it's up until round 20). It is of course subject to balance.

**Comment by Zet0rz** [comment]
![](https://thumbs.gfycat.com/SparklingUnimportantDalmatian-size_restricted.gif)
https://gfycat.com/SparklingUnimportantDalmatian

Buildable Tables can now be scripted in and very easily let you build anything (even power switches or PaP)! You simply create the table wherever you want, then do something like this:

``` Lua
local tbl = ents.Create("buildable_table")
tbl:AddValidCraft("shield", buildabletbl)
tbl:Spawn()
```

Where "buildabletbl" is a table of this format:

``` Lua
local buildabletbl = {
    model = "models/weapons/w_zombieshield.mdl",
    pos = Vector(0,0,60), -- (Relative to tables own pos)
    ang = Angle(0,180,90), -- (Relative too)
    parts = {
        ["shield1"] = {0,1}, -- Submaterials to "unhide" when this part is added
        ["shield2"] = {2}, -- id's are ItemCarry object IDs
        ["shield3"] = {3},
        -- You can have as many as you want
    },
    usefunc = function(ply) -- When it's completed and a player presses E (optional)
        if !ply:HasWeapon("nz_zombieshield") then
            ply:Give("nz_zombieshield")
        end
    end,
    text = function() -- Text to display when player is looking (after finished) (optional)
        if !ply:HasWeapon("nz_zombieshield") then
            return "Press E to pick up Zombie Shield"
        end
    end
}
```

What's left now is to make it non-instant (like revive, progress bar, probably use PaP arms viewmodel), and get the text working.

**Comment by minin43** [comment]
Yo paste this in the project wiki so I can use it, it looks awesome :D

**Comment by Zet0rz** [comment]
In the map script wiki pages I'll make an example/reference page with code snippets to copy-paste and modify, including a full buildable shield so you only have to edit the spawnpoint positions for the parts. It'll also have teleporters (if I get around to make that), maybe some fancy round modifiers.

**Comment by minin43** [comment]
You make this so easy for us, Zet.

**Comment by Zet0rz** [comment]
![](https://thumbs.gfycat.com/KeenImaginativeAlaskanhusky-size_restricted.gif)
(https://gfycat.com/KeenImaginativeAlaskanhusky)

Timed Use system allows for stuff similar to this. If any entity contains these functions:

``` Lua
function ENT:StartTimedUse(activator, caller, usetype, val) -- Replicated ENT:Use arguments
    -- Do stuff like give the nz_packapunch_arms weapon
    return 2 -- The time for the use to finish (progress bar) - 0 becomes instant use
end

function ENT:StopTimedUse(activator, caller, usetype, val)
    -- When the player cancels the use (by looking elsewhere or releasing E)
end

function ENT:FinishTimedUse(activator, caller, usetype, val)
    -- When the player has completed the time
    -- Make sure to re-check requirements checked in StartTimedUse in case they changed
end
```

Then this entity will no longer use it's own Use function, but instead these. You can call self:Use(a,b,c,d) in the FinishTimedUse() function to make use of it though (hence why the arguments are the same).
Map scripts can also use these simply by creating the functions on the entity

``` Lua
local ent = ents.Create("nz_script_prop")
ent.StartTimedUse = function(self, activator, caller, usetype, val)
    -- stuff
end
-- etc.
```

**Comment by Zet0rz** [comment]
Update to the content pack adds the WIP Zombie Shield so you can playtest the new config and map (which was also just added on the dev branch). Plan to speed up its deploy animation, and perhaps add damaged variants to it.

**Comment by func-octo** [comment]
Would those damaged variants be textures or moving the parts just a little to imply damage?

**Comment by Zet0rz** [comment]
I'm not sure. Any ideas? I can't really edit the textures as easily as they just use normal game textures of the fence and everything. The model is literally just an import of base HL2 models put together into 1 and given a bone to attach to players. I was thinking I could cut holes in the fence, or maybe bend the wood or something, but you don't really see alot in the first person view anyway ...

**Comment by func-octo** [comment]
Well, find some sounds that sound like something breaking, play those when it goes to the next damage phase. As for the visual damage, you could apply the burn marks from explosions onto the radiator or put bullet hole textures on it to signify damage. I think when it reaches the next damage phase, emit some sparks from it. I know it isn't a machine that would create sparks, but it would certainly alert the player to its damage.

**Comment by func-octo** [comment]
Actually, I have an idea that might work. You could add an LCD screen to the Shield in the middle right of the screen (in First-person) that displays the shield outlined in white, the background is black. When it reaches (1 of the 3 damage levels?) damaged states, you could make one of the parts turn red to signify the damage to the player while they have it out.

**Comment by minin43** [comment]
Wait, I'm confused, what do you need ideas for?

**Comment by func-octo** [comment]
@minin43 He needs ideas on how to make the shield look damaged in 3-4 stages.

**Comment by lolleko** [comment]
Change the color, or add damaged models.

**Comment by Lucius1202** [comment]
Each level of damage could remove an object from the shield.

**Comment by Zet0rz** [comment]
> you could apply the burn marks from explosions onto the radiator or put bullet hole textures on it to signify damage.

Burnt/bullet marks might work

> Change the color, or add damaged models.

That's obvious and would also be the best solution. I wanted to try bending some parts of it, but I don't know how it might affect the UV mapping. I could change color in-game too, make it go darker and darker.

> Each level of damage could remove an object from the shield.

There's 4 parts (subtextures) to the shield; The fence frame, the fence itself, the radiator, and the poles. I can remove (make them invisible) any of them, but in doing so the other parts will appear floating. If I make the fence get removed first (the outer part), it won't look like a shield at all.

> Well, find some sounds that sound like something breaking, play those when it goes to the next damage phase.

It has sounds every time it takes damage, and a bigger metal breaking sound when it breaks

**Comment by minin43** [comment]
Maybe make parts of the chain link invisible? As if the chain is starting to tear.

**Comment by Zet0rz** [comment]
> Maybe make parts of the chain link invisible? As if the chain is starting to tear.

That's what I was thinking, but it'd be in the mesh and not the texture (so just cut some holes).

**Comment by Zet0rz** [comment]
![](http://i.imgur.com/HI2dPxH.png)

![](http://i.imgur.com/8XSIbtR.png)

![](http://i.imgur.com/SBrbYtI.png)

Good?

**Comment by minin43** [comment]
A more drastic difference would be better. This looks good but it could look fantastic.

**Comment by Zet0rz** [comment]
Couldn't really change the fence itself, and bending the iron bars would be wierd as that's what you're holding on to

**Comment by minin43** [comment]
Oh I guess the more I look at it the more subtle differences I'm noticing.

But I guess I have to ask the question, do you realize what it looks like in-game?

**Comment by func-octo** [comment]
Could it be possible to make sections of the radiator fall off?

**Comment by Zet0rz** [comment]
![](http://i.imgur.com/qmVXUjT.jpg)

Should be the last piece of content the content pack needs, until I start working on the final boss which'll probably need a few custom sounds and models.

This looks good though! Random box spawnpoints will now take this model and stay visible in Survival :D

**Comment by minin43** [comment]
I forgot this existed. Well done!

**Comment by Zet0rz** [comment]
PaP camos might actually be a thing with string.find. I can exclude any materials including some words like hands or arms, and apply a camo to the rest:
https://gfycat.com/ClumsyRevolvingBunny

Might wanna randomize it slightly so that around half the materials are still normal. Looks better than a full gun camo.

Feedback on the camo used? That's just one of the fancy ones gmod comes with, as seen in the materials tool form Sandbox. Should it be a different one and/or should there be more than just 1?

**Comment by minin43** [comment]
Hmm, looks really good but how does it work?

**Comment by sh4d0w59** [comment]
How has this project been going?  Haven't heard from any of the devs in a while.

**Comment by isaiahtronik** [comment]
When I try to create maps in Nzombies I get tons of missing textures and errors even though I have both content packs installed
